bolt-sidecar/src/driver.rs

```diff
@@ -220,21 +220,26 @@ impl<C: StateFetcher, BLS: SignerBLS, ECDSA: SignerECDSA> SidecarDriver<C, BLS,
 
         // parse the request into constraints and sign them
         let slot = inclusion_request.slot;
-        let message = ConstraintsMessage::build(validator_pubkey, inclusion_request);
-        let signed_constraints = match self.constraint_signer.sign(&message.digest()).await {
-            Ok(signature) => SignedConstraints { message, signature },
-            Err(err) => {
-                error!(?err, "Failed to sign constraints");
-                let _ = response.send(Err(CommitmentError::Internal));
-                return;
-            }
-        };
 
-        // Track the number of transactions preconfirmed considering their type
-        signed_constraints.message.transactions.iter().for_each(|full_tx| {
-            ApiMetrics::increment_transactions_preconfirmed(full_tx.tx_type());
-        });
-        self.execution.add_constraint(slot, signed_constraints);
+        // NOTE: we iterate over the transactions in the request and generate a signed constraint for each one. This is because
+        // the transactions in the commitment request are not supposed to be treated as a relative-ordering bundle, but a batch
+        // with no ordering guarantees.
+        for tx in inclusion_request.txs {
+            let tx_type = tx.tx_type();
+            let message = ConstraintsMessage::from_transaction(validator_pubkey.clone(), slot, tx);
+
+            let signed_constraints = match self.constraint_signer.sign(&message.digest()).await {
+                Ok(signature) => SignedConstraints { message, signature },
+                Err(err) => {
+                    error!(?err, "Failed to sign constraints");
+                    let _ = response.send(Err(CommitmentError::Internal));
+                    return;
+                }
+            };
+
+            ApiMetrics::increment_transactions_preconfirmed(tx_type);
+            self.execution.add_constraint(slot, signed_constraints);
```

Comment:
Does this work if a transaction appears twice, first time in a commitment request with just one tx and the second time in a commitment with a batch of two transactions?
Because there is a risk otherwise we increment the nonce of balance of the user two times

Reply:
I created a quick test with a duplicate tx in a batch, which should and does fail with nonce too low. It's not exactly what you described but should cover it.